crmsh/cibquery.py

```diff
@@ -0,0 +1,26 @@
+"""utilities for parsing CIB xml"""
+import dataclasses
+import typing
+
+import lxml.etree
+
+
+@dataclasses.dataclass(frozen=True)
+class ResourceAgent:
+    m_class: str
+    m_provider: typing.Optional[str]
+    m_type: str
+
+
+def get_configured_resource_agents(cib: lxml.etree.Element) -> typing.Set[ResourceAgent]:
+    return set(
+        ResourceAgent(e.get('class'), e.get('provider'), e.get('type'))
+        for e in cib.xpath('/cib/configuration/resources//primitive')
+    )
+
+
+def has_primitive_filesystem_with_fstype(cib: lxml.etree.Element, fstype: str) -> bool:
```

Comment:
For commonly use purpose, I do think this function can be renamed more commonly, like `has_primitive_ra_with`, with parameters like type and nvpair's name and value

Reply:
I do know xml queries can be more flexible, just like JAXB in java. But a complete, user-friendly and reliable implementation needs too much work and is impossible to be included in crmsh. I incline to implement some routines only when we really need to call it from many places. Do you have some real use cases?